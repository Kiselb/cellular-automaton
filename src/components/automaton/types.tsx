export type AutomatonDescriptionBit = 0 | 1;
export type AutomatonDescription = {
  id: number;
  name: string;
  born: [
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit
  ];
  save: [
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit,
    AutomatonDescriptionBit
  ];
};
export const AutomatonsList: AutomatonDescription[] = [
  {
    id: 1,
    name: "Вольфрама-Паккарда, класс 2 (B018/S018)",
    //     0  1  2  3  4  5  6  7  8
    born: [1, 1, 0, 0, 0, 0, 0, 0, 1],
    save: [1, 1, 0, 0, 0, 0, 0, 0, 1],
  },
  {
    id: 2,
    name: "Вольфрама-Паккарда (B0578/S045)",
    //     0  1  2  3  4  5  6  7  8
    born: [1, 0, 0, 0, 0, 1, 0, 1, 1],
    save: [1, 0, 0, 0, 1, 1, 0, 0, 0],
  },
  {
    id: 3,
    name: "Вольфрама-Паккарда (B0578/S12456)",
    //     0  1  2  3  4  5  6  7  8
    born: [1, 0, 0, 0, 0, 1, 0, 1, 1],
    save: [0, 1, 1, 0, 1, 1, 1, 0, 0],
  },
  {
    id: 4,
    name: "Вольфрама-Паккарда (B1/S012345678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 1, 0, 0, 0, 0, 0, 0, 0],
    save: [1, 1, 1, 1, 1, 1, 1, 1, 1],
  },
  {
    id: 5,
    name: "Gnarl (B1/S1)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 1, 0, 0, 0, 0, 0, 0, 0],
    save: [0, 1, 0, 0, 0, 0, 0, 0, 0],
  },
  {
    id: 6,
    name: "Gnarl (B123567/S0238)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 1, 1, 1, 0, 1, 1, 1, 0],
    save: [1, 0, 1, 1, 0, 0, 0, 0, 1],
  },
  {
    id: 7,
    name: "Репликатор (B13456/S01356)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 1, 0, 1, 1, 1, 1, 0, 0],
    save: [1, 1, 0, 1, 0, 1, 1, 0, 0],
  },
  {
    id: 8,
    name: "Репликатор (B135/S135)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 1, 0, 1, 0, 1, 0, 0, 0],
    save: [0, 1, 0, 1, 0, 1, 0, 0, 0],
  },
  {
    id: 9,
    name: "Репликатор (B1357/S1357)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 1, 0, 1, 0, 1, 0, 1, 0],
    save: [0, 1, 0, 1, 0, 1, 0, 1, 0],
  },
  {
    id: 10,
    name: "Вольфрама-Паккарда (B137/S45678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 1, 0, 1, 0, 0, 0, 1, 0],
    save: [0, 0, 0, 0, 1, 1, 1, 1, 1],
  },
  {
    id: 11,
    name: "Вольфрама-Паккарда, класс 3 (B236/S0468)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 1, 1, 0, 0, 1, 0, 0],
    save: [1, 0, 0, 0, 1, 0, 1, 0, 1],
  },
  {
    id: 12,
    name: "Вольфрама-Паккарда (B257/S27)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 1, 0, 0, 1, 0, 1, 0],
    save: [0, 0, 1, 0, 0, 0, 0, 1, 0],
  },
  {
    id: 13,
    name: "Бессмертная жизнь (B3/S012345678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 0, 0, 0],
    save: [1, 1, 1, 1, 1, 1, 1, 1, 1],
  },
  {
    id: 14,
    name: "Maze (B3/S12345)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 0, 0, 0],
    save: [0, 1, 1, 1, 1, 1, 0, 0, 0],
  },
  {
    id: 15,
    name: "Жизнь (B3/S23)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 0, 0, 0],
    save: [0, 0, 1, 1, 0, 0, 0, 0, 0],
  },
  {
    id: 16,
    name: "Автомат Вольфрама-Паккарда, класс 2 (B3/S234)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 0, 0, 0],
    save: [0, 0, 1, 1, 1, 0, 0, 0, 0],
  },
  {
    id: 17,
    name: "Коралл (B3/S45678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 0, 0, 0],
    save: [0, 0, 0, 0, 1, 1, 1, 1, 1],
  },
  {
    id: 18,
    name: "Коралл (B34/S03456)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 1, 0, 0, 0, 0],
    save: [1, 0, 0, 1, 1, 1, 1, 0, 0],
  },
  {
    id: 19,
    name: "Жизнь 34 (B34/S34)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 1, 0, 0, 0, 0],
    save: [0, 0, 0, 1, 1, 0, 0, 0, 0],
  },
  {
    id: 20,
    name: "Ассимиляция (B345/S4567)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 1, 1, 0, 0, 0],
    save: [0, 0, 0, 0, 1, 1, 1, 1, 0],
  },
  {
    id: 21,
    name: "Долгая жизнь (B345/S5)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 1, 1, 0, 0, 0],
    save: [0, 0, 0, 0, 0, 1, 0, 0, 0],
  },
  {
    id: 22,
    name: "Диамёба (B35678/5678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 1, 1, 1, 1],
    save: [0, 0, 0, 0, 0, 1, 1, 1, 1],
  },
  {
    id: 23,
    name: "Амёба (B357/S1358)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 1, 0, 1, 0],
    save: [0, 1, 0, 1, 0, 1, 0, 0, 1],
  },
  {
    id: 24,
    name: "Псевдожизнь (B357/S238)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 1, 0, 1, 0],
    save: [0, 0, 1, 1, 0, 0, 0, 0, 1],
  },
  {
    id: 25,
    name: "2х2 (B36/S125)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 1, 0, 0],
    save: [0, 1, 1, 0, 0, 1, 0, 0, 0],
  },
  {
    id: 26,
    name: "Гламурная жизнь (B36/S23)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 1, 0, 0],
    save: [0, 0, 1, 1, 0, 0, 0, 0, 0],
  },
  {
    id: 27,
    name: "Stains (B367/S2346)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 1, 1, 0],
    save: [0, 0, 1, 1, 1, 0, 1, 0, 0],
  },
  {
    id: 28,
    name: "Stains (B3678/S235678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 1, 1, 1],
    save: [0, 0, 1, 1, 0, 1, 1, 1, 1],
  },
  {
    id: 29,
    name: "День и Ночь (B3678/S34678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 1, 1, 1],
    save: [0, 0, 0, 1, 1, 0, 1, 1, 1],
  },
  {
    id: 30,
    name: "Движение (B368/S245)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 1, 0, 1],
    save: [0, 0, 1, 0, 1, 1, 0, 0, 0],
  },
  {
    id: 31,
    name: "Автомат Вольфрама-Паккарда (B378/S012345678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 0, 1, 1],
    save: [1, 1, 1, 1, 1, 1, 1, 1, 1],
  },
  {
    id: 32,
    name: "Коагуляция (B378/S235678)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 1, 0, 0, 0, 1, 1],
    save: [0, 0, 1, 1, 0, 1, 1, 1, 1],
  },
  {
    id: 33,
    name: "Города за стенами (B45678/S2345)",
    //     0  1  2  3  4  5  6  7  8
    born: [0, 0, 0, 0, 1, 1, 1, 1, 1],
    save: [0, 0, 1, 1, 1, 1, 0, 0, 0],
  },
];

export type Props = {
  defAutomaton: number;
  onAutomatonChange: (params: AutomatonDescription) => void;
  probe?: ({}) => void;
};

export { Automaton } from "./Automaton";
