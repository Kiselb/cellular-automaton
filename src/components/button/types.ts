import { ButtonCaption } from "../../domain/types";

type ButtonAction = () => void;

export type Props = {
  status: boolean;
  caption: ButtonCaption;
  onAction: ButtonAction;
  testId: string;
  probe?: ({}) => void;
};

export { Button } from "./Button";
