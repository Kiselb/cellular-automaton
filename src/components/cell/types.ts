export type Params = {
  generation: number;
  row: number;
  col: number;
  colorGammaMin?: string;
  colorGammaMax?: string;
  colorEmpty: string;
  colorCell: string;
};

export type Props = Omit<Params, "colorCell"> & {
  onClick: (cell: Params) => void;
  probe?: ({}) => void;
};

export type CalcColor = (
  generation: number,
  colorScaleMin?: string,
  colorScaleMax?: string
) => string;

export { Cell, calcCellColor } from "./Cell";
