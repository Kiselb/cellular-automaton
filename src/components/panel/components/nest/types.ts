import React from "react";

export type Props = {
  key: string;
  row: number;
  col: number;
  children: React.ReactNode;
};

export { Nest } from "./Nest";
