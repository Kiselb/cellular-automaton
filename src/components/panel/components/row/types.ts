import React from "react";

export type Props = {
  key: string;
  row: number;
  children: React.ReactNode;
};

export { Row } from "./Row";
