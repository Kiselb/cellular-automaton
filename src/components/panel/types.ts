import { Params } from "../cell/types";

export type Props = {
  data: number[][];
  onChange: (params: Params) => void;
  probe?: ({}) => void;
};

export { Panel } from "./Panel";
