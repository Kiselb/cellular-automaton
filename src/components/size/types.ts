export type SizeHandler = (size: number) => void;

export type Props = {
  onSizeChange: SizeHandler;
  minSize: number;
  maxSize: number;
  defSize: number;
  testId: string;
  probe?: ({}) => void;
};

export { Size } from "./Size";
