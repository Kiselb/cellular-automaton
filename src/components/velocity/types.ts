export type VelocityItem = {
  delay: number;
  name: string;
};
export type VelocityHandler = (delay: number) => void;
export type Props = {
  onVelocityChange: VelocityHandler;
  defVelocity: number;
  probe?: ({}) => void;
};

export const VelocitySet: VelocityItem[] = [
  { delay: 1000, name: "Медленно" },
  { delay: 2000, name: "Очень медленно" },
  { delay: 100, name: "Быстро" },
  { delay: 10, name: "Очень быстро" },
  { delay: 500, name: "Нормально" },
];

export { Velocity } from "./Velocity";
