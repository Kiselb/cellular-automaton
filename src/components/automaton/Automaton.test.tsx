import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";

import "@testing-library/jest-dom";
import { Automaton } from "./types";
import { DEF_AUTOMATON } from "../../domain/defaults";
import { AutomatonDescription, AutomatonsList } from "./types";

it("Automaton renders correctly", () => {
  const { asFragment } = render(
    <Automaton
      defAutomaton={DEF_AUTOMATON}
      onAutomatonChange={() => console.log("OK")}
    />
  );
  screen.debug();
  expect(asFragment()).toMatchSnapshot();
});
it("Automaton has style", () => {
  render(
    <Automaton
      defAutomaton={DEF_AUTOMATON}
      onAutomatonChange={() => console.log("OK")}
    />
  );
  const div = screen.getByTestId("automaton");
  expect(div).toHaveClass("automaton");
});
it("Default automaton defined correctly", () => {
  const automaton: AutomatonDescription = AutomatonsList.filter(
    (automaton) => automaton.id === DEF_AUTOMATON
  )[0];
  expect(automaton.id).toEqual(DEF_AUTOMATON);
});
it("Automaton fire event on select", () => {
  const probe = jest.fn();
  render(<Automaton defAutomaton={DEF_AUTOMATON} onAutomatonChange={probe} />);
  const control = screen.getByTestId("automaton");
  fireEvent.change(control);
  expect(probe).toBeCalled();
});
