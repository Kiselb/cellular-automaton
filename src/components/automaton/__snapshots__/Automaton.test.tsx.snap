// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Automaton renders correctly 1`] = `
<DocumentFragment>
  <select
    class="automaton"
    data-testid="automaton"
  >
    <option
      value="25"
    >
      2х2 (B36/S125)
    </option>
    <option
      value="5"
    >
      Gnarl (B1/S1)
    </option>
    <option
      value="6"
    >
      Gnarl (B123567/S0238)
    </option>
    <option
      value="14"
    >
      Maze (B3/S12345)
    </option>
    <option
      value="27"
    >
      Stains (B367/S2346)
    </option>
    <option
      value="28"
    >
      Stains (B3678/S235678)
    </option>
    <option
      value="31"
    >
      Автомат Вольфрама-Паккарда (B378/S012345678)
    </option>
    <option
      value="16"
    >
      Автомат Вольфрама-Паккарда, класс 2 (B3/S234)
    </option>
    <option
      value="23"
    >
      Амёба (B357/S1358)
    </option>
    <option
      value="20"
    >
      Ассимиляция (B345/S4567)
    </option>
    <option
      value="13"
    >
      Бессмертная жизнь (B3/S012345678)
    </option>
    <option
      value="2"
    >
      Вольфрама-Паккарда (B0578/S045)
    </option>
    <option
      value="3"
    >
      Вольфрама-Паккарда (B0578/S12456)
    </option>
    <option
      value="4"
    >
      Вольфрама-Паккарда (B1/S012345678)
    </option>
    <option
      value="10"
    >
      Вольфрама-Паккарда (B137/S45678)
    </option>
    <option
      value="12"
    >
      Вольфрама-Паккарда (B257/S27)
    </option>
    <option
      value="1"
    >
      Вольфрама-Паккарда, класс 2 (B018/S018)
    </option>
    <option
      value="11"
    >
      Вольфрама-Паккарда, класс 3 (B236/S0468)
    </option>
    <option
      value="26"
    >
      Гламурная жизнь (B36/S23)
    </option>
    <option
      value="33"
    >
      Города за стенами (B45678/S2345)
    </option>
    <option
      value="30"
    >
      Движение (B368/S245)
    </option>
    <option
      value="29"
    >
      День и Ночь (B3678/S34678)
    </option>
    <option
      value="22"
    >
      Диамёба (B35678/5678)
    </option>
    <option
      value="21"
    >
      Долгая жизнь (B345/S5)
    </option>
    <option
      selected=""
      value="15"
    >
      Жизнь (B3/S23)
    </option>
    <option
      value="19"
    >
      Жизнь 34 (B34/S34)
    </option>
    <option
      value="32"
    >
      Коагуляция (B378/S235678)
    </option>
    <option
      value="17"
    >
      Коралл (B3/S45678)
    </option>
    <option
      value="18"
    >
      Коралл (B34/S03456)
    </option>
    <option
      value="24"
    >
      Псевдожизнь (B357/S238)
    </option>
    <option
      value="7"
    >
      Репликатор (B13456/S01356)
    </option>
    <option
      value="8"
    >
      Репликатор (B135/S135)
    </option>
    <option
      value="9"
    >
      Репликатор (B1357/S1357)
    </option>
  </select>
</DocumentFragment>
`;
